### DISCUSSION TITLE
Extension module can't find shared library from same Meson project

### DISCUSSION THREAD

**Discussion by sampotter** [mesonbuild/meson-python#565] [General]
I develop a numerical library in C called [butterfly](https://github.com/sampotter/butterfly) which is built using Meson and have more recently been using meson-python to build a thin wrapper for it written in Cython. By default, the C library is built as a shared library and the Python module just consumes this shared library as a dependency in the natural way in Meson. This has worked fine for testing when have I run scripts etc "locally" from `builddir` in the hackiest way possible.

I've just recently tried running `meson install` to see whether `import butterfly` will work. It doesn't. When I run `meson install` I see something like:
```
Installing subdir /home/sfp/Dropbox/Research/Butterfly/butterfly/include/bf to /usr/local/include/bf
Installing /home/sfp/Dropbox/Research/Butterfly/butterfly/include/bf/backends.h to /usr/local/include/bf
...
... many irrelevant lines ...
...
Installing /home/sfp/Dropbox/Research/Butterfly/butterfly/include/bf/vectors.h to /usr/local/include/bf
Installing libbutterfly.so to /usr/local/lib64
Installing wrappers/python/butterfly.cpython-311-x86_64-linux-gnu.so to /usr/local/lib64/python3.11/site-packages/
Installing /home/sfp/Dropbox/Research/Butterfly/butterfly/builddir/meson-private/butterfly.pc to /usr/local/lib64/pkgconfig
```
So, the shared library and Python module seem to get put somewhere reasonable. But, when I fire up Python and try to import the shared library, it can't be found:
```
[sfp@fedora ~]$ python
Python 3.11.7 (main, Dec 18 2023, 00:00:00) [GCC 13.2.1 20231011 (Red Hat 13.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import butterfly
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: libbutterfly.so: cannot open shared object file: No such file or directory
```
Any help would be appreciated. 

I should say that I'm very green when it comes to packaging. I would like to make this work but it's not clear to me what tool in this complicated process is even responsible.

Thanks!

**Comment by rgommers** [comment]
Hi @sampotter, thanks for the question. This has come up multiple times before, so I guess it's really time we write some docs that address this use case. (TODO: https://mesonbuild.com/meson-python/explanations/internal-dependencies.html)

First a few general pointers:
- You say you are running `meson install`, which means you are not using `meson-python` at all I think. `meson-python` is a "build backend" in Python packaging terminology, which means it gets used under the hood when building an sdist or wheel via commands like `python -m build` or `pip install .` (locally) or `pip install butterfly` (when `butterfly` is on PyPI)
  - If you avoid `meson-python` and indeed install with `meson setup`/`compile`/`install`, then you should be able to make things work mostly, you just won't get Python packaging metadata and you cannot install with `pip` & co.
- shared libraries are difficult for Python packaging; in particular it's important to realize that (a) `pip`/`build` are still able to install things into `<prefix>/lib/` but that that is strongly discouraged, and (b) project-external shared libraries are not handled by default at all, it needs an external tool like `auditwheel` to bundle them into a wheel. 
  - For _project-internal_ libraries things should work through `meson-python` though, except for the documented limitation on Windows (currently): https://mesonbuild.com/meson-python/reference/limitations.html#shared-libraries-on-windows
  - Also import: there is no fixed relationship in directory structure between `<prefix>/lib/` and Python's `site-packages`, so if you install via a wheel it's difficult to keep an RPATH working

To say more, it'd be useful to know what exactly you're aiming for here:
1. Installing with `meson install` only, and making your Python wrapper work that way
2. Using `meson-python` to produce wheels _including_ the shared library
3. Using `meson-python` to produce wheels _excluding_ the shared library, but only the Cython wrapper and relying on the user or a Linux distro to install that shared library separately.

And if (2) or (3): do you need it to work on Windows? And do you intend to distribute via PyPI, or only for users cloning your repo?

**Comment by sampotter** [comment]
Hi @rgommers. Thanks for the speedy and very educational response. I truly appreciate it! I did see the linked stub in the documentation and assumed it would be pertinent.

Re: your first general pointer: you can see the depth of my misunderstanding. :-) I had assumed that all Python-related functionality was  captured by meson-python, but it's clear now that `.extension_module` etc are provided by Meson's [Python](https://mesonbuild.com/Python-module.html) module. Are there long term plans for meson-python to merge into Meson proper?

And thanks for the explanation regarding shared libraries. I have seen this come up enough times in other contexts that I figured it would be kind of nasty to resolve. If I could trouble you: do you have any recommendations for something to read about RPATHs? I'm not very familiar with them or how they work and they seem central to this business.

To explain what I'm after:

1. The core C library is where all important functionality lives. Since it is a numerical library, the long-term goal is to provide thin wrappers for the main scientific languages (Python, Julia, MATLAB, etc... even C++). 
2. I would like to make it so that whoever is using it can install it in the way that seems easiest and most convenient for them: i.e., clone + install following README, `pip install ...`, etc. I don't care what these build steps look like. If using `meson install` to install everything makes the most sense, fine; or if running meson *and* pip is necessary, also fine. My goal is to make things work consistently and reliably. Happy to add an extra step or two to the process if it means things will work consistently.

Since I would like someone to be able to just `pip install butterfly` (or for another Python project to depend on `butterfly`), it sounds like PyPI + wheels is mandatory. And in this case, since it's discouraged for pip to install a dependency itself, the shared library should be bundled, or the C library should even just be built statically and linked...

Butterfly also depends on several other numerical libraries: OpenBLAS, ARPACK, UMFPACK, GSL, etc, which get picked up either using Meson's `find_library` or `dependency`. It sounds like this may also not play nicely with Python packaging...

**Reply by dnicolodi** [reply]
> Are there long term plans for meson-python to merge into Meson proper?

The Meson maintainers are open to the idea of integrating the functionality provided by meson-python directly in Meson. However, meson-python requires a few libraries to work, and there is the desire of do not have Meson depend on nothing else than the Python standard library. On the other hand, there is no advantage in merging meson-python into Meson: the introspection data exported by Meson is enough for meson-python to do its job. Therefore, I don't think meson-python will be merged into Meson any time soon.

> do you have any recommendations for something to read about RPATHs? I'm not very familiar with them or how they work and they seem central to this business.

Here is a simplified explanation. Executables and libraries (Python modules are in essence shared libraries) can link dynamically do symbols exported by shared libraries. When the code is about to be executed, the dynamic linker needs to look up the shared library and resolve the symbols to their actual location in memory. The linker looks up the library name in a number of filesystems paths. Most often, the shared library is found in one of the handful of location where the dynamic linker is configured to look. Sometimes, shared library need to be installed in locations outside these search paths. These additional lookup locations can be specified as `RPATH` fields int the ELF executable file format. `man ld.so` for more details. Generally, you don't need to care about `RPATH`s: shared libraries are installed in one of the paths where the dynamic linker searches for them. The Mach-O executable file format has an equivalent of the `RPATH` fields. The executable file format used by Windows does not.

Things get more complicated when distributing Python extension modules in wheels (as you do when you want to distribute precompiled binaries). The wheel format is structured to allow to install Python packages in virtual Python environments. This clashes with the requirement of installing shared libraries in system wide locations where the dynamic linker looks for them. One solution to this problem (implemented transparently by meson-python, on UNIX platforms and macOS) is to install the shared libraries in a non standard location (compatible with the Python virtual environment location) and set an `RPATH` (or equivalent, on macOS) to instruct the dynamic linker to look there for shared libraries when loading the Python extension modules. Because Windows lacks an equivalent concept to `RPATH`, this does not work there. There are other tricks that can be used to still distribute Python extension modules linking to DLLs on Windows.

> Since I would like someone to be able to just pip install butterfly (or for another Python project to depend on butterfly), it sounds like PyPI + wheels is mandatory. 

Technically, you can distribute only a source package for your Python package, in the form of an `sdist` and have `pip install butterfly` compile the extension modules on the users' system. This however, requires the users to have a compiler toolchain and all dependencies installed on their system. Depending on the expected users of your library, this may be a very high bar to clear.

> And in this case, since it's discouraged for pip to install a dependency itself

"discouraged" is not the right word: pip cannot install dependencies that are not distributed as Python packages. 

> Butterfly also depends on several other numerical libraries: OpenBLAS, ARPACK, UMFPACK, GSL, etc, which get picked up either using Meson's find_library or dependency. It sounds like this may also not play nicely with Python packaging...

If you have only one extension module, there is no advantage in dynamically linking to these libraries: the size of the wheel with the dynamically linked extension module plus all the shared libraries is actually likely to be larger than then wheel with the statically linked extension module. If you go for static linking, building and distributing a Python wheel is very easy. If you have more than one extension module yuo may want to link dynamically. SciPy solved this problem for many of the same libraries and @rgommers may provide some details on how they do it.

All this, however, has nothing to do with the problem you reported in your initial post: when you execute `meson install` the shared library should have been installed in one of the global paths searched by the dynamic linker and the issue you report should not occur. What do `/sbin/ldconfig -v 2>/dev/null | grep -v ^$'\t'` and `ldd /usr/local/lib64/python3.11/site-packages/butterfly.cpython-311-x86_64-linux-gnu.so` return?

**Reply by sampotter** [reply]
Thanks for the tremendously useful information. I haven't had a chance to digest it and research further yet, but will be following up on this soon.

In the mean time, here's the output of those two commands:
```
[sfp@fedora ~]$ ldconfig -v 2>/dev/null | grep -v ^$'\t'
/usr/lib64/R/lib: (from /etc/ld.so.conf.d/R-x86_64.conf:1)
/usr/lib64/atlas: (from /etc/ld.so.conf.d/atlas-x86_64.conf:1)
/usr/lib64/dyninst: (from /etc/ld.so.conf.d/dyninst-x86_64.conf:1)
/usr/lib64/ghc-9.2.6/lib/x86_64-linux-ghc-9.2.6: (from /etc/ld.so.conf.d/ghc.conf:1)
/usr/lib64/iscsi: (from /etc/ld.so.conf.d/libiscsi-x86_64.conf:1)
/usr/lib64/llvm13/lib: (from /etc/ld.so.conf.d/llvm13-x86_64.conf:1)
/usr/lib64/llvm15/lib: (from /etc/ld.so.conf.d/llvm15-x86_64.conf:1)
/usr/i686-w64-mingw32/lib: (from /etc/ld.so.conf.d/mingw32-hostlib.conf:1)
/usr/x86_64-w64-mingw32/lib: (from /etc/ld.so.conf.d/mingw64-hostlib.conf:1)
/usr/lib64/pipewire-0.3/jack: (from /etc/ld.so.conf.d/pipewire-jack-x86_64.conf:1)
/usr/lib64/qt-3.3/lib: (from /etc/ld.so.conf.d/qt-x86_64.conf:1)
/usr/lib64/qt5-qtwebengine-freeworld: (from /etc/ld.so.conf.d/qt5-qtwebengine-freeworld-x86_64.conf:1)
/usr/lib64/tcl8.6: (from /etc/ld.so.conf.d/tix-x86_64.conf:1)
/lib: (from <builtin>:0)
/lib64: (from <builtin>:0)
```
and:
```
[sfp@fedora ~]$ ldd /usr/local/lib64/python3.11/site-packages/butterfly.cpython-311-x86_64-linux-gnu.so 
	linux-vdso.so.1 (0x00007ffe5d153000)
	libbutterfly.so => not found
	libpython3.11.so.1.0 => /lib64/libpython3.11.so.1.0 (0x00007f7ef3c00000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f7ef4155000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f7ef3b1f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f7ef4403000)
```

**Reply by dnicolodi** [reply]
Looking at the output from `ldconfig` and at the `meson install` log you posted previously, the problem is that Meson installed the shared library in `/usr/local/lib64/` but this path is not in the list of the paths where the dynamic linker looks for shared libraries. It is customary to install software that is not packaged by the distribution into `/usr/local/`, and for the dynamic linker to be configured to look into `/usr/local/lib/` or `/usr/local/lib64/`. I don't know why your distribution is not doing that.

What Linux distribution are you using? You may ask on the support channel for your distribution for guidance.

Or you can simply drop a file in `/etc/ld.so.conf.d/` with a line containing `/usr/local/lib64` and issue `ldconfig` as the root user.

**Reply by sampotter** [reply]
I'm running Fedora 38. It appears that having `/usr/local/lib` and `/usr/local/lib64` globally in the dynamic linker path might be discouraged on Fedora for some reason. Searching around, I was able to turn up some comments related to this, but couldn't find a good explanation. I would assume Meson would get this right by default since Fedora is a very widely used distribution...

**Reply by sampotter** [reply]
For now I've fixed this by passing `--prefix=/usr` when running `meson setup ...`.

**Reply by dnicolodi** [reply]
Which path is used on Fedora for locally installed software that is not installed as a distribution package? 

> I would assume Meson would get this right by default since Fedora is a very widely used distribution...

I don't understand what you mean. `/usr/local/` is the default installation prefix used by Meson on Unix-like systems. If you install in this prefix, libraries go in `/usr/local/lib/` or `/usr/local/lib64/`. If you want to use another installation prefix, just specify it. There isn't anything Meson is doing wrong or that Meson could be doing differently.

**Reply by sampotter** [reply]
It does appear that Fedora uses `/usr/local` for software that isn't a distribution package.

It also looks like Fedora omits `/usr/local/lib` or `/usr/local/lib64` on purpose, while many other distributions don't.

Because Meson's (reasonable) default prefix is `/usr/local`, it appears to me I will need to include a notice with my library directed to Fedora users that they need to either pass `--prefix=/usr` (maybe problematic?) or add something to `/etc/ld.so.conf.d` (annoying).
